### PR TITLE
Bj 07

### PR DESCRIPTION
**Pull request by BryanJusto**
## Issue No.

N/A

## Description

Monster.harmlessEncounter -> Monster.encounterDescription
GetEncounterDescription moved to it's own java file
QuickEnvDetails is now implemented in the code

Added
HazardEnvDetails, Generates a hazard description
Automated testing for 60% of the codebase

Removed
Unused lines of code.
old way of having harmlessDescription

## Testing

Manually checked the output from the expected result to determine that it was consistent.

## Test Cases

Automated the testing of the randomEncounter.java file, I bruteforced multiple iterations to ensure no out of bounds errors were occurring. Further tests are needed to be implemented 

## Double check:
- [Y] Change was tested manually
- [Y] New code has automated tests

### CODE REVIEW

File: src/main/java/com/pathfinderGenerator/app/generator/RandomEncounter.java [hunk 1]
```diff
@@ -139,58 +139,45 @@ private List<Monster> creatureRandomEncounter(RandomEncounterObj randomEncounter
...
         int rnd = rollDice.rollDice(20);
-        if(rnd >= 21){
-//            System.out.println("rnd = " + rnd);
-            rnd = rnd%20;
-        }
 
         //it has rolled higher than the static DC lets see what we get now
         if(rnd > randomEncounterObj1.getDc()) {
             int rndHazard = rollDice.rollDice(10);
-            switch (rndHazard) {
```
Comment (keithfrancisb) [inline src/main/java/com/pathfinderGenerator/app/generator/RandomEncounter.java:169]: based on how we have rolling dice defined, I think we don't need the plus 1 here. according to [this](https://www.geeksforgeeks.org/java-math-random-method-examples/) I think we don't need it since the min is always 1, we'll always be subtracting and adding 1.

Comment (keithfrancisb) [reply src/main/java/com/pathfinderGenerator/app/generator/RandomEncounter.java:169]: also I don't think adding a 1 should live here. adding the 1 should live in the function `rollDice`.

File: src/main/java/com/pathfinderGenerator/app/generator/StyleGenerator.java [hunk 2]
```diff
@@ -518,7 +518,7 @@ public Monster createMonsterQuery(String name) throws SQLException, ClassNotFoun
     }
 
     private ResultSet queryDatabase(String sqlString) throws ClassNotFoundException, SQLException {
-        System.out.println("*********** query = " + sqlString);
```
Comment (keithfrancisb) [inline src/main/java/com/pathfinderGenerator/app/generator/StyleGenerator.java:521]: do we want to comment this back in?

File: src/main/java/com/pathfinderGenerator/app/generator/randomEncounterTools/GetEncounterDescription.java [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+
+    RollDice rollDice = new RollDice();
+
+    HazardEnvDetails hazardEnvDetails = new HazardEnvDetails();
+
+    public String getHarmlessEncounterDescription(String terrain){
+        int rnd = rollDice.rollDice(5);
+        String sense = "";
+        switch (rnd) {
+            case 0: sense = "Sights"; break;
```
Comment (keithfrancisb) [inline src/main/java/com/pathfinderGenerator/app/generator/randomEncounterTools/GetEncounterDescription.java:22]: it shouldn't be possible to roll a 0 right?

File: src/main/java/com/pathfinderGenerator/app/generator/RandomEncounter.java (hunk 1 above)
Comment (keithfrancisb) [reply src/main/java/com/pathfinderGenerator/app/generator/RandomEncounter.java:169]: it seems like the `rollDice` function is also used in places where 0 can be the min. it might be ideal to adjust the function or add a separate function to handle cases where the min is 0. ideally maybe with a name that's something not related to dice since the min is always 1.

File: src/main/java/com/pathfinderGenerator/app/generator/randomEncounterTools/GetEncounterDescription.java [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+                break;
+            case "Mountain":
+                description = (hazardEnvDetails.mountain()[rnd]);
+                break;
+            default:
+                description = (hazardEnvDetails.plains()[rnd]);
+                break;
+        }
+
+        return description;
```
Comment (keithfrancisb) [inline src/main/java/com/pathfinderGenerator/app/generator/randomEncounterTools/GetEncounterDescription.java:84]: I think it makes more sense for the rng to live in `HazardEnvDetails`. the `aquatic` and similar functions can just randomly generate the description.

File: src/main/java/com/pathfinderGenerator/app/generator/randomEncounterTools/HazardEnvDetails.java [hunk 5]
```diff
@@ -0,0 +1,65 @@
+package com.pathfinderGenerator.app.generator.randomEncounterTools;
+
+public class HazardEnvDetails {
+
+    public String[] aquatic() {
+        String[] Forest = new String[5];
```
Comment (keithfrancisb) [inline src/main/java/com/pathfinderGenerator/app/generator/randomEncounterTools/HazardEnvDetails.java:6]: similar to most sting lists here, the descriptions don't pertain to forests lol. also its common practice for variable names to start with a lowercase and list variables names to be plural. I suggest just using `descriptions` for all of these lists.

File: src/main/java/com/pathfinderGenerator/app/generator/randomEncounterTools/LevelBasedDc.java [hunk 6]
```diff
@@ -0,0 +1,10 @@
+package com.pathfinderGenerator.app.generator.randomEncounterTools;
+
+public class LevelBasedDc {
+    public int levelBasedDc(int level){
+        int startingDc = 14;
+        int diffAdder = level/3;
```
Comment (keithfrancisb) [inline src/main/java/com/pathfinderGenerator/app/generator/randomEncounterTools/LevelBasedDc.java:6]: do we not need to do any rounding?

File: src/main/java/com/pathfinderGenerator/app/object/Monster.java [hunk 7]
```diff
@@ -91,5 +91,5 @@ public String toString() {
     // Denotes the type of the encounter. If null it's probably a monster
     private String encType;
     // TODO: figure out type checking and subclasses and separate this class out
```
Comment (keithfrancisb) [inline src/main/java/com/pathfinderGenerator/app/object/Monster.java:93]: let's create an issue for this todo and link it here

Comment (keithfrancisb) [reply src/main/java/com/pathfinderGenerator/app/object/Monster.java:93]: unless this is already done then lets remove the comment

**Review by keithfrancisb — COMMENTED** [review]
left some comments. do you think you can get the code coverage up to at least 80% in this PR?

File: src/main/java/com/pathfinderGenerator/app/generator/randomEncounterTools/GetEncounterDescription.java (hunk 3 above)
Comment (BryanJusto) [reply src/main/java/com/pathfinderGenerator/app/generator/randomEncounterTools/GetEncounterDescription.java:22]: it is possible to roll a 1 here

Comment (keithfrancisb) [reply src/main/java/com/pathfinderGenerator/app/generator/randomEncounterTools/GetEncounterDescription.java:22]: you mean a 0? I might have added this comment before adding the comment about the rollDice function utilizing a min factor

File: src/main/java/com/pathfinderGenerator/app/generator/RandomEncounter.java (hunk 1 above)
Comment (BryanJusto) [reply src/main/java/com/pathfinderGenerator/app/generator/RandomEncounter.java:169]: > also I don't think adding a 1 should live here. adding the 1 should live in the function `rollDice`.

we use the rolldice to rng from an array, as well as rng a switch statement, so if we add a 1 in the rollDice.java then we need to subtract a -1 when reading into an array, which is the same issue, we are just moving it around



> it seems like the `rollDice` function is also used in places where 0 can be the min. it might be ideal to adjust the function or add a separate function to handle cases where the min is 0. ideally maybe with a name that's something not related to dice since the min is always 1.

this would add unneeded bloat to the code, since the two are functionally the same we are just adding a 1 to what would be the exact same coed anyway. 

> based on how we have rolling dice defined, I think we don't need the plus 1 here. according to [this](https://www.geeksforgeeks.org/java-math-random-method-examples/) I think we don't need it since the min is always 1, we'll always be subtracting and adding 1.

int rand = (int)(Math.random() * range) + min; they are hardcoding a +1 in the example you provided, it always goes from 0 -> maxval

File: src/main/java/com/pathfinderGenerator/app/generator/randomEncounterTools/GetEncounterDescription.java (hunk 4 above)
Comment (BryanJusto) [reply src/main/java/com/pathfinderGenerator/app/generator/randomEncounterTools/GetEncounterDescription.java:84]: > I think it makes more sense for the rng to live in `HazardEnvDetails`. the `aquatic` and similar functions can just randomly generate the description.

the main reason why I did not put it there is to reduce the amount of lines of code (e.g. 1 line in `GetEncounterDescription` vs 5 in `HazardEnvDetails`) 

the other main reason I choose against this is so we can reuse the `HazardEnvDetails`, in the case we want to be able to select a specific a specific hazard, if we make it choose the RNG inside and only return the hazard chosen by rng instead of returning the entire list and choosing randomly from that list.

would there be a bigger benefit to having the RNG choose within the `HazardEnvDetails` vs just leaving it in `GetEncounterDescription` ?

File: src/main/java/com/pathfinderGenerator/app/generator/randomEncounterTools/HazardEnvDetails.java (hunk 5 above)
Comment (BryanJusto) [reply src/main/java/com/pathfinderGenerator/app/generator/randomEncounterTools/HazardEnvDetails.java:6]: > similar to most sting lists here, the descriptions don't pertain to forests lol. also its common practice for variable names to start with a lowercase and list variables names to be plural. I suggest just using `descriptions` for all of these lists.

yeah this was a lazy carry over, the original code had it like that I just changed a bit of it around Forest is an old carry over

File: src/main/java/com/pathfinderGenerator/app/generator/randomEncounterTools/LevelBasedDc.java (hunk 6 above)
Comment (BryanJusto) [reply src/main/java/com/pathfinderGenerator/app/generator/randomEncounterTools/LevelBasedDc.java:6]: > do we not need to do any rounding?

No, int value will always round down e.g. 3.33 -> 3 and that is what we want. We could do Math.Floor() but it would be the same outcome and would probably slow the calculation down slightly.

File: src/main/java/com/pathfinderGenerator/app/object/Monster.java (hunk 7 above)
Comment (BryanJusto) [reply src/main/java/com/pathfinderGenerator/app/object/Monster.java:93]: Not going to lie that is 2 years old I have to try and remember what we were trying to say with this todo

File: src/main/java/com/pathfinderGenerator/app/generator/RandomEncounter.java [hunk 8]
```diff
@@ -139,58 +139,45 @@ private List<Monster> creatureRandomEncounter(RandomEncounterObj randomEncounter
...
-        descMap.put("Forest", "diverse flora and fauna, natural remedies, plentiful game");
-        descMap.put("Mountain", "minerals (including gold, silver, and gemstones)");
-        descMap.put("Plain", "useful plants (flax or cotton)");
-        descMap.put("Swamp", " materials useful for medicine or poison");
-        return descMap;
-    }
-
     public List<Monster> rndEncGenerator(String randomEncounterObj, int partySize, int partyLevel, List<String> sourceList){
 
         List<Monster> userEncounter = new LinkedList<>();
```
[the remark this answers is not in the document]
Comment (BryanJusto) [reply src/main/java/com/pathfinderGenerator/app/generator/RandomEncounter.java:144]: > I might be missing it, but is there a reason why this is a list?

In the event of a Combat encounter it can return multiple creatures, so if it is a non-combat encounter it returns a list of 1, otherwise it returns a list of multiple creatures.

File: src/main/java/com/pathfinderGenerator/app/generator/RandomEncounter.java (hunk 1 above)
Comment (keithfrancisb) [reply src/main/java/com/pathfinderGenerator/app/generator/RandomEncounter.java:169]: right. so it seems there are cases when the min is 0 and there are other cases where the min is 1. so it might be ideal to have two functions. one for when the min is 1 and another for when the min is 0. another thing we could do is make the function optionally take in a min and use that in the rng formula, otherwise default to 1 since you can't get a 0 when "rolling a die".

File: src/main/java/com/pathfinderGenerator/app/generator/RandomEncounter.java [hunk 9]
```diff
@@ -139,58 +139,45 @@ private List<Monster> creatureRandomEncounter(RandomEncounterObj randomEncounter
...
                 case 7:
-//                    hazard?
-//                    System.out.println("A hazard has occurred, this could be like a cliff, or toxic mushrooms / " +
-//                            "corrupted land, and is not combat based. ");
+                    //Hazards
+                    Monster encounter2 = new Monster();
+                    encounter2.setEncType("hazard");
+                    encounter2.setEncounterDescription(getEncounterDescription.getHazardEncounterDescription(
+                            randomEncounterObj1.getTerrain()) + " Base DC: " + levelBasedDc.levelBasedDc(partyLevel));
+                    userEncounter = Collections.singletonList(encounter2);
```
[the remark this answers is not in the document]
Comment (BryanJusto) [reply src/main/java/com/pathfinderGenerator/app/generator/RandomEncounter.java:172]: > 

That causes an error as Monster encounter is already declared as part of the scope 
image

File: src/main/java/com/pathfinderGenerator/app/generator/randomEncounterTools/GetEncounterDescription.java (hunk 4 above)
Comment (keithfrancisb) [reply src/main/java/com/pathfinderGenerator/app/generator/randomEncounterTools/GetEncounterDescription.java:84]: logically, I think it just makes more sense. like it doesn't need to know that the function is generating the descriptions randomly. we could make the functions take in an optional argument to specify which descriptions we want them to return if we don't want it to generate them randomly. imo, I think its better to favor readability (and understandability?) over less lines of code.

**Comment by BryanJusto** [comment]
> left some comments. do you think you can get the code coverage up to at least 80% in this PR?

I do not, since that would require to make test code for the entire backend, which is not in scope for this pr, this pr is only for the `RandomEnounter` and that code coverage is up to 80% already, another pr should be used for the other parts of the backend, we probably don't want to just have me working on all of the test code in one push, along side functional changes to the code.

tl;dr this code coverage is just for the new code, which is already up to 80%, future pr will work on the other aspects of the backend. 

image

File: src/main/java/com/pathfinderGenerator/app/generator/randomEncounterTools/GetEncounterDescription.java (hunk 4 above)
Comment (keithfrancisb) [reply src/main/java/com/pathfinderGenerator/app/generator/randomEncounterTools/GetEncounterDescription.java:84]: feel free to keep it, but I personally think it just makes more sense in the other class.

File: src/main/java/com/pathfinderGenerator/app/generator/RandomEncounter.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (keithfrancisb) [reply src/main/java/com/pathfinderGenerator/app/generator/RandomEncounter.java:144]: so right now, there's just no logic yet for returning a list of multiple creatures?

[the remark this answers is not in the document]
Comment (keithfrancisb) [reply src/main/java/com/pathfinderGenerator/app/generator/RandomEncounter.java:144]: why not use `ArrayList` instead of a `LinkedList`?

File: src/main/java/com/pathfinderGenerator/app/generator/randomEncounterTools/LevelBasedDc.java (hunk 6 above)
Comment (keithfrancisb) [reply src/main/java/com/pathfinderGenerator/app/generator/randomEncounterTools/LevelBasedDc.java:6]: ah right I miss the int part

File: src/main/java/com/pathfinderGenerator/app/generator/RandomEncounter.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (BryanJusto) [reply src/main/java/com/pathfinderGenerator/app/generator/RandomEncounter.java:144]: > so right now, there's just no logic yet for returning a list of multiple creatures?

I apologize, I think I am confused, we are using a list to return multiple creatures, thats why we are using a list.


> why not use `ArrayList` instead of a `LinkedList`?

would these two not function the same?

File: src/main/java/com/pathfinderGenerator/app/generator/RandomEncounter.java (hunk 1 above)
Comment (BryanJusto) [reply src/main/java/com/pathfinderGenerator/app/generator/RandomEncounter.java:169]: > right. so it seems there are cases when the min is 0 and there are other cases where the min is 1. so it might be ideal to have two functions. one for when the min is 1 and another for when the min is 0. another thing we could do is make the function optionally take in a min and use that in the rng formula, otherwise default to 1 since you can't get a 0 when "rolling a die".

I am going to pushback on the writing two functions that are the exact same code but one has a 0 as a minimum and one has a 1 as a minimum as it will require more testing to ensure it is working, and bloat the codebase with unneeded functions.

As for the optional input I can work on that, shouldn't be too much of a code change to implement such a thing.

File: src/main/java/com/pathfinderGenerator/app/generator/StyleGenerator.java (hunk 2 above)
Comment (BryanJusto) [reply src/main/java/com/pathfinderGenerator/app/generator/StyleGenerator.java:521]: I mean I can re-add it, it's just makes the System.out.println be really hard to read when you get a giant wall of text

File: src/main/java/com/pathfinderGenerator/app/generator/RandomEncounter.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (keithfrancisb) [reply src/main/java/com/pathfinderGenerator/app/generator/RandomEncounter.java:144]: it would, but I think it's just better to default to `ArrayList`.

File: src/main/java/com/pathfinderGenerator/app/generator/StyleGenerator.java (hunk 2 above)
Comment (keithfrancisb) [reply src/main/java/com/pathfinderGenerator/app/generator/StyleGenerator.java:521]: let's just remove it then. no need to commit it in the project.

File: src/main/java/com/pathfinderGenerator/app/object/Monster.java (hunk 7 above)
Comment (keithfrancisb) [reply src/main/java/com/pathfinderGenerator/app/object/Monster.java:93]: let's just remove the TODO comment then. we can add it back if we need to.

**Review by keithfrancisb — APPROVED** [review]
